packages/core/src/config/parse.ts

```diff
@@ -1655,10 +1688,78 @@ const parseAndValidateChugSplashConfig = async (
     contracts: {},
   }
 
+  // Resolve all the contracts upfront so they can be used handle contract references regardless or ordering.
+  // We also cache the compiler output, constructor args, and other artifacts so we don't have to re-read them later when we complete the parsing
   const contracts = {}
+  const cachedCompilerOutput = {}
+  const cachedConstructorArgs = {}
+  const cachedArtifacts = {}
   for (const [referenceName, userContractConfig] of Object.entries(
     userConfig.contracts
   )) {
+    const { externalProxy, kind, constructorArgs } = userContractConfig
+
+    // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
+    // used by ChugSplash.
+    const proxy =
+      externalProxy ||
+      getDefaultProxyAddress(userConfig.options.projectName, referenceName)
+
+    const { output } = readBuildInfo(artifactPaths[referenceName].buildInfoPath)
+    cachedCompilerOutput[referenceName] = output
+
+    const artifact = readContractArtifact(
+      artifactPaths[referenceName].contractArtifactPath,
+      integration
+    )
+    cachedArtifacts[referenceName] = artifact
+    const { abi, bytecode, sourceName, contractName } = artifact
+
+    const { args, valid: validArgs } = parseAndValidateConstructorArgs(
+      constructorArgs ?? {},
+      referenceName,
+      output.contracts[sourceName][contractName].abi,
+      cre
+    )
+    cachedConstructorArgs[referenceName] = args
+
+    if (kind === 'no-proxy') {
+      if (validArgs !== true) {
+        // If the constructor arugments are invalid, then we use address zero as the contract address.
+        // This is ok because the deployment wont actually happen due to the invalid constructor arugments.
+        // Setting it to the AddressZero just allows the rest of the parsing and validation logic to complete without errors.
+        contracts[referenceName] = ethers.constants.AddressZero
+      } else {
+        const creationCodeWithConstructorArgs =
+          getCreationCodeWithConstructorArgs(bytecode, args, referenceName, abi)
+
+        contracts[referenceName] = getContractAddress(
+          parsedConfig.options.projectName,
+          creationCodeWithConstructorArgs
+        )
+      }
+    } else {
+      contracts[referenceName] = proxy
+    }
```

Comment:
Here we call the `parseAndValidateConstructorArgs` function to perform validation on the constructor arguments. I've reintroduced the valid return value pattern for this specific case because valid constructor arguments are necessary to calculate the address for non-proxied contracts. If we pass invalid arguments to the `getContractAddress` function then we get an esoteric error and the program crashes. 

This is kind of like some of the other cases where some validation failing causes the rest of the logic to not work correctly. But in those cases, we just throw a useful error and exit the program entirely when the relevant validation fails to avoid outputting weird errors. In this case, 99% of the time if the constructor argument validation fails it's fine because the arguments are only used for non-proxied contracts. So I didn't want to just exit the program if the constructor args aren't valid. Instead, I decided to just track if the constructor args are invalid, and use the zero address if so. This allows the rest of the parsing logic to work fine, and the program will exit at the end due to the constructor arguments being invalid.

Reply:
IMO we should throw errors for invalid constructor args up-front instead of following this pattern.

> 99% of the time if the constructor argument validation fails it's fine because the arguments are only used for non-proxied contracts.

Constructor args are used for immutable variables in proxied contracts, so I expect they'll be used very frequently

Also, it seems like an anti-pattern to rely on logic outside of `parseAndValidateConstructorArgs` to throw an error for invalid constructor args. It's not clear to me where exactly this error would eventually be thrown in our codebase

Reply:
>> IMO we should throw errors for invalid constructor args up-front instead of following this pattern.
We do, the error is caused by invalid constructor arguments and the issue is logged during the `parseAndValidateConstructorArgs` function. As a result, the program will cleanly exit after the rest of the parsing logic is finished running.

>> Constructor args are used for immutable variables in proxied contracts, so I expect they'll be used very frequently
Not within the parsing logic which is the section of the code that I am referring to. In the parsing logic, we use the constructor arguments only to calculate the address of stateless immutable contracts. If the constructor arguments are invalid, then no other part of the codebase is relevant because the program will end after the parsing logic is completed.

>> it seems like an anti-pattern to rely on logic outside of `parseAndValidateConstructorArgs` to throw an error for invalid constructor args
I agree, that's exactly what I am trying to avoid with this logic. If we call `getCreationCodeWithConstructorArgs` with invalid arguments that an esoteric error will be thrown due to the arguments not matching the ABI. This logic avoids that by not calculating the address if and only if the arguments are valid. 

An alternative solution would be to throw an error and completely stop the program anytime an invalid constructor argument is detected in `parseAndValidateConstructorArgs`, but that doesn't seem like a great idea to me for two reasons:
1. This issue is not relevant to proxied contracts which means we'd be stopping execution early for no reason the vast majority of the time. 
2. We've explicitly set the goal of trying to output as many errors to the user at once as possible like a compiler. So I don't think it makes sense to deviate from that as long as it doesn't require large amounts of refactoring.

Reply:
Got it, I understand now. IMO we should still throw a separate set of errors for invalid constructor arguments. Here's my reasoning:
1. The code in its current state is non-intuitive to me. It took me a little while to understand why exactly we're doing it this way. That may just be a "me problem", but I wouldn't be surprised if future readers struggle to understand what's going on too
2. The issue of invalid constructor arguments will become a more frequent occurrence if/when we support stateful non-proxied contracts.
3. Compilers don't output all classes of errors at once. If you mess around with a Solidity contract, you'll see what I mean. Generally speaking, we should strive to output as many errors at once, but I think the tradeoff here isn't worth it.

Happy to hop on a call if you don't agree